### PR TITLE
Checkout / Sales Rules / Get rid of redundant DB query

### LINKED ISSUE #29453: [Issue] Checkout / Sales Rules / Get rid of redundant DB query
This issue is automatically created based on existing pull request: magento/magento2#29376: Checkout / Sales Rules / Get rid of redundant DB query

---------
**Preconditions:**

> - Loading data in the loop?
> - No, thank you!

By replacing direct SQL query with Data Provider that was already used in this Validator class allowed me to drastically reduce the cost of checkout placing process.

#### Local environment

20 products in cart

image

#### Custom project

22 products in cart

image

### Fixed Issues (if relevant)
Performance of Checkout

**Steps to reproduce:**

1. Install sample data
2. Create Catalog rule based on category
3. Add plenty of products (15  - inf.)
4. Place an order.

**Actual Result:**

**Expected Result:**

Checkout should work faster by 20 - 30% 

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by lbajsarowicz**
By replacing direct SQL query with Data Provider that was already used in this Validator class allowed me to drastically reduce the cost of checkout placing process.

#### Local environment

20 products in cart

image

#### Custom project

22 products in cart

image

### Fixed Issues (if relevant)
Performance of Checkout

### Manual testing scenarios (*)
1. Install sample data
2. Create Catalog rule based on category
3. Add plenty of products (15  - inf.)
4. Place an order.

Enjoy the pleasure of the checkout faster by 20 - 30% (and/or) send me a good whiskey

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29453: Checkout / Sales Rules / Get rid of redundant DB query

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@slavvka @lenaorobei is there an ability to run performance against 2.4-develop and for this PR? Would be good to check the diff

**Comment by IvanChepurnyi** [comment]
Nice performance optimization @lbajsarowicz, would be great also to remove the singular query with preloaded category ids into this service by using afterload even on product collection in the shopping cart.

**Comment by lbajsarowicz** [comment]
@magento run Static Tests, Unit Tests

These failed due to implementation testing. SVC has to be approved by Magento crew.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @lbajsarowicz,
Could you implement following suggestion? https://github.com/magento/magento2/pull/29376#issuecomment-668468327

**Comment by lbajsarowicz** [comment]
@ihor-sviziev That is the subject for another PR.
With the delays caused by Product Owners from Adobe, I'd prefer to do it separately.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]
✔ Approved.

Failing SVC test looks false positive

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7950](https://jira.corp.magento.com/browse/ENGCOM-7950) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Install Magento with sample data
2. Create Catalog rule based on category
3. Add plenty of products (30 in my case)
4. Place an order.

**Local environment**

**Before:** :heavy_multiplication_x: 



**After:** :heavy_check_mark:
 30 products in cart

Screenshot from 2020-08-06 12-43-06

**Comment by engcom-Charlie** [comment]
**Semantic Version Checker** failed with error 
>Magento\Rule\Model\Condition\Product\AbstractProduct::_getAvailableInCategories `[protected] Method return typing changed.`

It needs to approve.

File: app/code/Magento/Rule/Model/Condition/Product/AbstractProduct.php [hunk 1]
```diff
@@ -657,23 +658,11 @@ public function validateByEntityId($productId)
      * Retrieve category ids where product is available
      *
      * @param int $productId
-     * @return array
```
Comment (ihor-sviziev) [inline app/code/Magento/Rule/Model/Condition/Product/AbstractProduct.php]: @lbajsarowicz can we move it to separate PR in order not to wait for approval? This PR brings good performance improvement!

Comment (lbajsarowicz) [reply app/code/Magento/Rule/Model/Condition/Product/AbstractProduct.php]: Yes, we can 🤣

**Comment by lbajsarowicz** [comment]
@magento run all tests

I just rolled back PHPDoc clarification, to avoid Magento approval procedures hell.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7950](https://jira.corp.magento.com/browse/ENGCOM-7950) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Install Magento with sample data
2. Create Catalog rule based on category
3. Add plenty of products (50 in my case)
4. Place an order.

**Local environment**

**Before:** :heavy_multiplication_x: 




**After:** :heavy_check_mark:
50 products in cart

Screenshot from 2020-08-07 14-29-24

**Comment by engcom-Alfa** [comment]
@magento create issue

**Pull request merged**